Tests/Packet++Test/Tests/FtpTests.cpp

```diff
@@ -0,0 +1,177 @@
+					 "Authentication password.");
+	PTF_ASSERT_EQUAL(pcpp::FtpRequestLayer::getCommandAsString(pcpp::FtpRequestLayer::PASS), "PASS");
+
+	READ_FILE_AND_CREATE_PACKET(5, "PacketExamples/ftpIpv6Resp.dat");
+
+	pcpp::Packet ftpPacket5(&rawPacket5);
+	pcpp::FtpResponseLayer *ftpLayer5 = ftpPacket5.getLayerOfType<pcpp::FtpResponseLayer>();
+
+	PTF_ASSERT_NOT_NULL(ftpLayer5);
+	PTF_ASSERT_EQUAL(ftpLayer5->getStatusCode(), pcpp::FtpResponseLayer::COMMAND_NOT_IMPLEMENTED);
+	PTF_ASSERT_EQUAL(ftpLayer5->getStatusCodeString(), "502");
+	PTF_ASSERT_EQUAL(ftpLayer5->getStatusOption(), "Unknown command 'utf8'.");
+	PTF_ASSERT_EQUAL(ftpLayer5->toString(), "FTP Response: 502");
+	PTF_ASSERT_FALSE(ftpLayer5->isMultiLine());
+
+	PTF_ASSERT_EQUAL(pcpp::FtpResponseLayer::getStatusCodeAsString(pcpp::FtpResponseLayer::COMMAND_NOT_IMPLEMENTED),
+					 "Command not implemented");
+}
+
+PTF_TEST_CASE(FtpCraftingTests)
+{
+	timeval time;
+	gettimeofday(&time, NULL);
+
+	// Craft packets
+	READ_FILE_AND_CREATE_PACKET(1, "PacketExamples/ftpIpv4Req.dat");
+
+	pcpp::Packet ftpPacket1(&rawPacket1);
+	pcpp::Packet craftedPacket1;
+
+	pcpp::EthLayer ethLayer1(*ftpPacket1.getLayerOfType<pcpp::EthLayer>());
+	PTF_ASSERT_TRUE(craftedPacket1.addLayer(&ethLayer1));
+
+	pcpp::IPv4Layer ipv4Layer1(*ftpPacket1.getLayerOfType<pcpp::IPv4Layer>());
+	PTF_ASSERT_TRUE(craftedPacket1.addLayer(&ipv4Layer1));
+
+	pcpp::TcpLayer tcpLayer1(*ftpPacket1.getLayerOfType<pcpp::TcpLayer>());
+	PTF_ASSERT_TRUE(craftedPacket1.addLayer(&tcpLayer1));
+
+	pcpp::FtpRequestLayer ftpReqLayer1;
+	ftpReqLayer1.setCommand(pcpp::FtpRequestLayer::USER);
+	ftpReqLayer1.setCommandOption(std::string("csanders"));
+	PTF_ASSERT_TRUE(craftedPacket1.addLayer(&ftpReqLayer1));
+
+	PTF_ASSERT_EQUAL(bufferLength1, craftedPacket1.getRawPacket()->getRawDataLen());
+	PTF_ASSERT_BUF_COMPARE(buffer1, craftedPacket1.getRawPacket()->getRawData(), bufferLength1);
+
+	READ_FILE_AND_CREATE_PACKET(2, "PacketExamples/ftpIpv4RespHyphen.dat");
+
+	pcpp::Packet ftpPacket2(&rawPacket2);
+	pcpp::Packet craftedPacket2;
+
+	pcpp::EthLayer ethLayer2(*ftpPacket2.getLayerOfType<pcpp::EthLayer>());
+	PTF_ASSERT_TRUE(craftedPacket2.addLayer(&ethLayer2));
+
+	pcpp::IPv4Layer ipv4Layer2(*ftpPacket2.getLayerOfType<pcpp::IPv4Layer>());
+	PTF_ASSERT_TRUE(craftedPacket2.addLayer(&ipv4Layer2));
+
+	pcpp::TcpLayer tcpLayer2(*ftpPacket2.getLayerOfType<pcpp::TcpLayer>());
+	PTF_ASSERT_TRUE(craftedPacket2.addLayer(&tcpLayer2));
+
+	pcpp::FtpResponseLayer ftpReqLayer2;
+	ftpReqLayer2.setStatusCode(pcpp::FtpResponseLayer::SYSTEM_STATUS);
+	ftpReqLayer2.setStatusOption(
+		std::string("Extensions supported:\r\n CLNT\r\n MDTM\r\n PASV\r\n REST STREAM\r\n SIZE\r\n211 End."));
+	PTF_ASSERT_TRUE(craftedPacket2.addLayer(&ftpReqLayer2));
+
+	PTF_ASSERT_EQUAL(bufferLength2, craftedPacket2.getRawPacket()->getRawDataLen());
+	PTF_ASSERT_BUF_COMPARE(buffer2, craftedPacket2.getRawPacket()->getRawData(), bufferLength2);
+
+	// Modify existing packets
```

Comment:
Usually when modifying packets, one of the best ways to test is by saving it to raw data and check that we get what we expect. What you can do is:
- Write custom code to save the packet raw data to a pcap file
- Open it in Wireshark to make sure you get the expected result
- Convert the pcap file to `.dat` and save it in `PacketExamples/` folder
- Remove the code that saves the raw data to file
- Load the `.dat` file in the test
- Use `PTF_ASSERT_BUF_COMPARE` to compare the packet data with the content of the `.dat` file, like [here](https://github.com/seladb/PcapPlusPlus/blob/master/Tests/Packet%2B%2BTest/Tests/SipSdpTests.cpp#L624)

Reply:
I'll update it but I want to work on other comments first